### DISCUSSION TITLE
Index exceeds the number of array elements

### DISCUSSION THREAD

**Discussion by nargesmbg** [mtex-toolbox/mtex#1827] [Ask Anything]
Hello everyone, 

I did ebsd measurement on five different areas. and I'm trying to do grain sizing measurement but I receive following error! Error in Index exceeds the number of array elements. Index must not exceed
1266829.

Error in spatialDecomposition (line 41)
  D = D(1:size(X,1));

Error in EBSD/calcGrains (line 67)
[V,F,I_FD] = spatialDecomposition([ebsd.prop.x(:),
ebsd.prop.y(:)],ebsd.unitCell,varargin{:});

Error in untitled3 (line 81)
grains = calcGrains(ebsd);

can someone help me to solve the problem?

**Comment by bartvdbraak** [comment]
Hi @nargesmbg

I can help you with the error you're experiencing during the grain sizing measurement in MTEX. The error message suggests that the index used exceeds the number of elements in the array. Specifically, it states that the index must not exceed 1266829.

To troubleshoot this issue, let's take a look at the code snippet you provided. The error seems to occur in the `spatialDecomposition` function, specifically on line 41. In that line, there is an attempt to adjust the size of the array `D` to match the size of array `X`. However, it appears that the size mismatch is causing the error.

One possible solution is to ensure that the dimensions of `D` and `X` are compatible. You can try modifying the code on line 41 to ensure that `D` is resized correctly based on the size of `X`. For example, you can replace the line with:

```matlab
D = D(1:size(X, 1), :);
```

This modification will adjust the size of `D` based on the number of rows in `X`, while preserving the number of columns.

Additionally, it's worth checking the inputs you are providing to the `calcGrains` function on line 67. Ensure that the `ebsd` object and the optional arguments are correctly defined and match the expected format.

Please give these suggestions a try and let me know if the error persists or if you need further assistance. Providing more details about your code and the specific versions of MTEX and MATLAB you are using would also be helpful in diagnosing the issue.

Looking forward to hearing from you!

Best regards,
Bart

**Reply by nargesmbg** [reply]
hi Bart, 

thanks for your detailed explanation. as you suggested I tried to replace line 41 with: " D = D(1:size(X, 1), :);" in spatialDecomposition.m , unfortunately I received the following error message saying that I don't have permission to overwrite or change this file! here is the error message I receive: 

Unable to save spatialDecomposition.m 
C:\Program Files\MATLAB\mtex-5.9.0\mtex-5.9.0\EBSDAnalysis\@EBSD\private\spatialDecomposition.m (Access is denied) 


do you have any idea how should we solve this problem?

**Comment by kilir** [comment]
Hi,

is it correct to assume it is your merged dataset from yesterday?
If so, does it plot correctly? How did you merge it?

Cheers,
Rüdiger

**Comment by nargesmbg** [comment]
I'd like to also put the piece of code which merge the ebsd data collected from different locations on my sample and meant to calculate to grain size: 
```
%% Import the Data

% create an EBSD variable containing the data
ebsd1 = EBSD.load(fname1,CS,'interface','ang',...
  'convertEuler2SpatialReferenceFrame');
ebsd2 = EBSD.load(fname2,CS,'interface','ang',...
  'convertEuler2SpatialReferenceFrame');
ebsd3 = EBSD.load(fname3,CS,'interface','ang',...
  'convertEuler2SpatialReferenceFrame');
ebsd4 = EBSD.load(fname4,CS,'interface','ang',...
  'convertEuler2SpatialReferenceFrame');
ebsd5 = EBSD.load(fname5,CS,'interface','ang',...
  'convertEuler2SpatialReferenceFrame');

ebsd = [ebsd1, ebsd2, ebsd3, ebsd4, ebsd5];

%% calculate Grain 

ebsd = ebsd('indexed');

% reconstruct grains
grains= calcGrains(ebsd)

% plot the orientation data of the Forsterite phase
plot(ebsd('Ni'),ebsd('Ni').orientations)

% plot the grain boundary on top of it
hold on
plot(grains.boundary,'lineWidth',2)
hold off
%% histogram of grain siye over grain area
close all
histogram(grains.area)
xlabel('grain area')
ylabel('number of grains')
```

**Reply by kilir** [reply]
Are you sure that all your ebsd datasets have the right x,y coordinates with respect to another? If not, you need to shift each dataset appropriatly before concatenating.

Otherwise, if you just want grain size, compute grains for each dataset individually, eventually discard grains being cut by the map boundaries, and join the lists of grain sizes ( whatever measure you like) for your analysis.

Cheers,
Rüdiger

**Reply by nargesmbg** [reply]
hi Rüdiger, 

how should I check that all my ebsd datasets have right x,y coordinates with respect to another?

**Reply by kilir** [reply]
Does the merged dataset plit correctly?
